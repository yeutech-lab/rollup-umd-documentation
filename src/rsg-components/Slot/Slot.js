// Inspired by https://github.com/camwest/react-slot-fill

import React from 'react';
import PropTypes from 'prop-types';

export default function Slot({ name, active, onlyActive, className, props = {} }, { slots }) {
	const fills = slots[name];
	if (!fills) {
		throw new Error(`Slot "${name}" not found, available slots: ${Object.keys(slots).join(', ')}`);
	}

	const rendered = fills.map((Fill, index) => {
		// { id: 'pizza', render: ({ foo }) => <div>{foo}</div> }
		const { id, render } = Fill;
		if (id && render) {
			// Render only specified fill
			if (onlyActive && id !== active) {
				return null;
			}

			const { onClick } = props;
			props = {
				...props,
				name: id,
				// Set active prop to active fill
				active: active && id === active,
				// Pass fill ID to onClick event handler
				// eslint-disable-next-line react/prop-types
				onClick: onClick && ((...attrs) => onClick(id, ...attrs)),
			};

			Fill = render;
		}

		return <Fill key={index} {...props} />;
	});

	const filtered = rendered.filter(Boolean);
	if (filtered.length === 0) {
		return null;
	}

	return <div className={className}>{filtered}</div>;
}

Slot.propTypes = {
	name: PropTypes.string.isRequired,
	active: PropTypes.string,
	onlyActive: PropTypes.bool,
	props: PropTypes.object,
	className: PropTypes.string,
};
Slot.contextTypes = {
	slots: PropTypes.object.isRequired,
};
