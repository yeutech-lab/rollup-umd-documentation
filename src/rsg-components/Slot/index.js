export { default } from '../Slot/Slot';
