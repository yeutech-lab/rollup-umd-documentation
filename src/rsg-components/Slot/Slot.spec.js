import React from 'react';
import Slot from './Slot';

/* eslint-disable react/prop-types */

const Button = ({ onClick }) => <button onClick={onClick}>1</button>;
const Button2 = () => <button>2</button>;
const fillsWithIds = [
	{
		id: 'one',
		render: Button,
	},
	{
		id: 'two',
		render: Button2,
	},
];

it('should renderer slots and pass props', () => {
	const actual = shallow(<Slot name="slot" props={{ id: 'Pizza' }} />, {
		context: {
			slots: {
				slot: [Button, Button2],
			},
		},
	});

	expect(actual).toMatchSnapshot();
});

it('should renderer slots in id/render format', () => {
	const actual = shallow(<Slot name="slot" props={{ id: 'Pizza' }} />, {
		context: {
			slots: {
				slot: fillsWithIds,
			},
		},
	});

	expect(actual).toMatchSnapshot();
});

it('should pass active flag to active slot', () => {
	const actual = shallow(<Slot name="slot" active="two" />, {
		context: {
			slots: {
				slot: fillsWithIds,
			},
		},
	});

	expect(actual).toMatchSnapshot();
});

it('should renderer only active slot if onlyActive=true', () => {
	const actual = shallow(<Slot name="slot" active="two" onlyActive />, {
		context: {
			slots: {
				slot: fillsWithIds,
			},
		},
	});

	expect(actual).toMatchSnapshot();
});

it('should pass slot ID to onClick handler', () => {
	const onClick = jest.fn();
	const actual = mount(<Slot name="slot" props={{ onClick }} />, {
		context: {
			slots: {
				slot: fillsWithIds,
			},
		},
	});

	actual
		.find('button')
		.first()
		.simulate('click');

	expect(onClick).toBeCalledWith('one', expect.any(Object));
});

it('should return null if all slots render null', () => {
	const actual = render(<Slot name="slot" props={{ id: 'Pizza' }} />, {
		context: {
			slots: {
				slot: [() => null],
			},
		},
	});

	expect(actual.node).toBeFalsy();
});
