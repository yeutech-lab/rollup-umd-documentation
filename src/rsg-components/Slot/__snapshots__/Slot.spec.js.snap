// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should pass active flag to active slot 1`] = `
<div>
  <Button
    active={false}
    key="0"
    name="one"
  />
  <Button2
    active={true}
    key="1"
    name="two"
  />
</div>
`;

exports[`should renderer only active slot if onlyActive=true 1`] = `
<div>
  <Button2
    active={true}
    key="1"
    name="two"
  />
</div>
`;

exports[`should renderer slots and pass props 1`] = `
<div>
  <Button
    id="Pizza"
    key="0"
  />
  <Button2
    id="Pizza"
    key="1"
  />
</div>
`;

exports[`should renderer slots in id/render format 1`] = `
<div>
  <Button
    id="Pizza"
    key="0"
    name="one"
  />
  <Button2
    id="Pizza"
    key="1"
    name="two"
  />
</div>
`;
